File: 0041-first-missing-positive/0041-first-missing-positive.java
class Solution {
    public int firstMissingPositive(int[] nums) {
        Set<Integer> set = new HashSet<Integer>();
        final int n = nums.length;
        int max = 0;
        for(int i = 0; i < n; i++) {
            if(nums[i] < n + 1 && nums[i] > 0 && set.contains(nums[i]) == false) {
                if(nums[i] > max) max = nums[i];
                set.add(nums[i]);
            }
        }
        
        for(int ans = 1; ans <= max; ans++) {
            if(set.contains(ans) == false) {
                return ans;
            }
        }
        return max + 1;
    }
}

File: 0041-first-missing-positive/0041-first-missing-positive.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var firstMissingPositive = function(nums) {
    
    let set = new Set();
    let max = 0;
    for(let num of nums) {
        max = Math.max(max, num);
        if(num > 0 && set.has(num) == false) {
            set.add(num);
        }
    }
    //let max = (2 ** 31) - 1;
    for(let i = 1; i <= max; i++) {
        if(set.has(i) == false) {
            return i;
        }
    }
    return max + 1;
};